### PR TITLE
8258961: move some fields of SafePointNode from public to protected

### PR DESCRIPTION
**Pull request by navyxliu**
Declare VMStructs is a friend of SafePointNode so we can move some fields from public zone to protected zone. 
Clean up the forward declaration section in callnode.hpp. Lock/Unlock Nodes are subclasses of AbstractLockNode. 

/cc hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258961](https://bugs.openjdk.java.net/browse/JDK-8258961): move some fields of SafePointNode from public to protected

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1899/head:pull/1899`
`$ git checkout pull/1899`

### CODE REVIEW

File: src/hotspot/share/opto/callnode.hpp [hunk 1]
```diff
@@ -323,9 +325,16 @@ class JVMState : public ResourceObj {
...
 class SafePointNode : public MultiNode {
+  friend class GraphKit;
+  friend class JVMState;
   virtual bool           cmp( const Node &n ) const;
   virtual uint           size_of() const;       // Size is bigger
 
+protected:
+  JVMState* _jvms;      // Pointer to list of JVM State objects
+  void set_jvms(JVMState* s) {
+    _jvms = s;  // override const attribute in the accessor
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/callnode.hpp:336]: The comment should be adjusted since `_jvms` is not const anymore.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/vmStructs.cpp [hunk 2]
```diff
@@ -941,8 +941,6 @@ typedef HashtableEntry<InstanceKlass*, mtClass>  KlassHashtableEntry;
   c2_nonstatic_field(JVMState,                 _method,                                       ciMethod*)                             \
   c2_nonstatic_field(JVMState,                 _map,                                          SafePointNode*)                        \
                                                                                                                                      \
-  c2_nonstatic_field(SafePointNode,            _jvms,                                         JVMState* const)                       \
-                                                                                                                                     \
```
Comment (vnkozlov) [inline src/hotspot/share/runtime/vmStructs.cpp:945]: Don't remove it. It is referenced from HotSpot Servicability agent:
https://github.com/openjdk/jdk/blob/master/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/opto/SafePointNode.java#L45

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Also update last copyright year in changed files.

File: src/hotspot/share/runtime/vmStructs.cpp (hunk 2 above)
Comment (navyxliu) [reply src/hotspot/share/runtime/vmStructs.cpp]: Thank you for pointing out this.  I wasn't aware of this agent thing. 
It turns out I can neither move SafePointNode::_jvms to the protected zone nor take off its const qualifier. 
so I revert my changes except removing "virtual" for the SafePointNode::jvms().

File: src/hotspot/share/opto/callnode.hpp (hunk 1 above)
Comment (navyxliu) [reply src/hotspot/share/opto/callnode.hpp]: thank you for reviewing this.   Vladimir helped me to understand the hotspot Servicability agent thing, which needs to access this field. I have to revert this change.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by navyxliu** [comment]
I investigate the 3 failures on x86 host. they have been fixed in JDK-8258703 (committed yesterday). 
Presumably, the regression tests are clear.  /integrate.

**Review by TobiHartmann — APPROVED** [review]

File: src/hotspot/share/runtime/vmStructs.cpp (hunk 2 above)
Comment (navyxliu) [reply src/hotspot/share/runtime/vmStructs.cpp]: I reverted files which need to update copyright years.

**Comment by TobiHartmann** [comment]
@navyxliu There is a title mismatch between the PR and the JBS issue that you need to fix before integration.

**Comment by navyxliu** [comment]
yes, I did modify the JBS title. I found it's hard to hide out

**Comment by navyxliu** [comment]
I found Node::jvms() is virtual, so it's useless to remove keyword "virtual" in SafePointNode::jvms(). I need another revision to fix it.  

Paul asked me to hide away those public fields into protected. I discover a straightforward solution --  just declare friend class VMStructs. Other classes do that.

**Comment by navyxliu** [comment]
I took back the devirtual attempt because it's not worth it.   there're 3 nodes override Node::jvms(). 
non-virtual jvms() looks more expensive to access vtable. 

```
JVMState* Node::jvms() const {
  if (is_SafePoint()) {
    return as_SafePoint()->jvms();
  } else if (is_MachSafePoint()) {
    return as_MachSafePoint()->jvms();
  } else if (is_MachHalt()) {
   return as_MachHalt()->jvms();
  }
  return nullptr;
}
```

**Comment by navyxliu** [comment]
hello, @TobiHartmann , 
Could you take a look at this PR again?  I've corrected title and description to reflect what this patch does.

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me but @vnkozlov should also review this again.

**Review by vnkozlov — APPROVED** [review]
Looks good.

Latest version passed hs-tier1-3 testing.

**Comment by navyxliu** [comment]
Thanks for all reviewers.

**Pull request closed without merging**